docs/decisions/0001-lti-extensions-plugin.rst

```diff
@@ -0,0 +1,116 @@
+the configuration storage.
+
+.. _`Merge LTI 1.3 support to master`: https://github.com/edx/xblock-lti-consumer/pull/82
+.. _`Add support for LTI embeds in course tabs and elsewhere`: https://github.com/edx/xblock-lti-consumer/pull/77
+
+
+A few actions that need to be taken to achieve the desired architecture are outlined below:
+
+Centralized configuration for LTI launches
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use Django models to store LTI related configuration and launch data for all LTI integrations.
+This will provide a consistent configuration store throughout the platform and a single point of
+maintenance and development for LTI related features (good since the LTI Specification is well defined).
+
+Additionaly this would include configurations UI on Studio (to enable setup by instructors) and
+also through Django Admin (enable for larger contexts: programs or platform-wide).
+
+
+New LTI Python API
+~~~~~~~~~~~~~~~~~~
+
+A new Python API needs to be developed to provide LTI support platform wide.
+This API should provide methods to configure, modify and perform LTI launches
+from multiple platform contexts (Django, XBlocks).
+
+A centralized launch & configuration API will make LTI integration consistent
+across all platform's contexts.
+
+Change folder structure and rename repository
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To better reflect the new repo usage, rename the repository to "openedx-lti-consumer" and
+change the folder structure.
+
+Here is the new proposed folder structure:
+```
+# Django plugabble app main folder
+lti_consumer/
+
+# Python APIs - following `Inter App APIs`_ guidelines
+lti_consumer/api.py
+
+# LTI specific logic (no storage)
+lti_consumer/lti_1p1
+lti_consumer/lti_1p3
+
+# LTI Configuration, launch and services storage
+lti_consumer/models/
+
+# Pluggable extensions points that can be used in the platform
+lti_consumer/extensions/
+lti_consumer/extensions/xblock.py  # To replace current XBlock LTI Consumer, can use course import/export hooks to change settings
+lti_consumer/extensions/django.py  # To easily enable LTI embeds anywhere in the platform
+lti_consumer/extensions/course_tab.py  # Inherit Django extension above and add helpers
+                                       # to use course context for a few use cases (forum tab, course embed)
+```
+
+:: _`Inter App APIs`: https://github.com/edx/edx-platform/blob/master/docs/decisions/0002-inter-app-apis.rst
+
+Tech Debt
+=========
+
+The current LTI implementation stores launch configuration and user data in XBlock Settings fields.
+We'll need a transition phase and a migration path before fully deprecating the old configuration storage.
+
+The transition phase can allow configuring LTI integration both locally (on the XBlock) and using the
+generic configuration.
+
+
+Consequences
+============
+
+* LTI configuration will be independent from courses, but hooks can be used to
+  import/export data to the LTI configuration store.
```

Comment:
Is this saying that when one imports/exports a course, the LTI configuration will import/export with it? I'm not sure if that's a good thing or a bad thing; for sharing courses within an org/instance, it makes sense, but for sharing courses across orgs it'd be better to omit the LTI credentials.

Reply:
Yeah, that makes more sense from a security perspective.
I think we can omit the LTI credentials and display validation errors in the blocks and pages using it.

In a course context, we can display errors in the blocks and also on the LTI configuration UI (something like: `your course is using **configuration-123** but that isn't set up`).